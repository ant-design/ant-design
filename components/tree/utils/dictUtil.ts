import type { DataNode, Key } from 'rc-tree/lib/interface';
import { fillFieldNames } from 'rc-tree/lib/utils/treeUtil';

import type { TreeProps } from '../Tree';

const RECORD_NONE = 0;
const RECORD_START = 1;
const RECORD_END = 2;

type Record = typeof RECORD_NONE | typeof RECORD_START | typeof RECORD_END;

type FieldNames = TreeProps['fieldNames'];

function traverseNodesKey(
  treeData: DataNode[],
  callback: (key: Key | number | null, node: DataNode) => boolean,
  fieldNames: Required<NonNullable<FieldNames>>,
) {
  const { key: fieldKey, children: fieldChildren } = fieldNames;

  function processNode(dataNode: DataNode & FieldNames[keyof FieldNames]) {
    const key = dataNode[fieldKey];
    const children = dataNode[fieldChildren];
    if (callback(key, dataNode) !== false) {
      traverseNodesKey(children || [], callback, fieldNames);
    }
  }

  treeData.forEach(processNode as any);
}

/** 计算选中范围，只考虑expanded情况以优化性能 */
export function calcRangeKeys({
  treeData,
  expandedKeys,
  startKey,
  endKey,
  fieldNames,
}: {
  treeData: DataNode[];
  expandedKeys: Key[];
  startKey?: Key;
  endKey?: Key;
  fieldNames?: FieldNames;
}): Key[] {
  const keys: Key[] = [];
  let record: Record = RECORD_NONE;

  if (startKey && startKey === endKey) {
    return [startKey];
  }
  if (!startKey || !endKey) {
    return [];
  }

  function matchKey(key: Key) {
    return key === startKey || key === endKey;
  }

  traverseNodesKey(
    treeData,
    (key) => {
      if (record === RECORD_END) {
        return false;
      }

      if (matchKey(key as any)) {
        // Match test
        keys.push(key as any);

        if (record === RECORD_NONE) {
          record = RECORD_START;
        } else if (record === RECORD_START) {
          record = RECORD_END;
          return false;
        }
      } else if (record === RECORD_START) {
        // Append selection
        keys.push(key as any);
      }
      return expandedKeys.includes(key as any);
    },
    fillFieldNames(fieldNames),
  );

  return keys;
}

export function convertDirectoryKeysToNodes(
  treeData: DataNode[],
  keys: Key[],
  fieldNames?: FieldNames,
) {
  const restKeys: Key[] = [...keys];
  const nodes: DataNode[] = [];
  traverseNodesKey(
    treeData,
    (key, node) => {
      const index = restKeys.indexOf(key as any);
      if (index !== -1) {
        nodes.push(node);
        restKeys.splice(index, 1);
      }

      return !!restKeys.length;
    },
    fillFieldNames(fieldNames),
  );
  return nodes;
}
