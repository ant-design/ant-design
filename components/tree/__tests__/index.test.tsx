import React from 'react';

import { render, screen } from '../../../tests/utils';
import Tree from '../index';
import type { AntTreeNodeProps } from '../Tree';

const { TreeNode } = Tree;

describe('Tree', () => {
  it('icon and switcherIcon of Tree with showLine should render correctly', () => {
    const { asFragment } = render(
      <Tree showLine showIcon>
        <TreeNode icon="icon" switcherIcon="switcherIcon" key="0-0">
          <TreeNode icon="icon" switcherIcon="switcherIcon" key="0-0-0" />
          <TreeNode switcherIcon="switcherIcon" key="0-0-1" />
          <TreeNode icon="icon" key="0-0-2" />
          <TreeNode key="0-0-3" />
        </TreeNode>
        <TreeNode switcherIcon="switcherIcon" key="0-1">
          <TreeNode icon="icon" switcherIcon="switcherIcon" key="0-0-0" />
          <TreeNode switcherIcon="switcherIcon" key="0-0-1" />
          <TreeNode icon="icon" key="0-0-2" />
          <TreeNode key="0-0-3" />
        </TreeNode>
        <TreeNode key="0-2">
          <TreeNode icon="icon" switcherIcon="switcherIcon" key="0-0-0" />
          <TreeNode switcherIcon="switcherIcon" key="0-0-1" />
          <TreeNode icon="icon" key="0-0-2" />
          <TreeNode key="0-0-3" />
        </TreeNode>
      </Tree>,
    );
    expect(asFragment().firstChild).toMatchSnapshot();
  });

  it('switcherIcon in Tree should not render at leaf nodes', () => {
    const { container } = render(
      <Tree switcherIcon={<i className="switcherIcon" />} defaultExpandAll>
        <TreeNode icon="icon">
          <TreeNode title="node1" icon="icon" key="0-0-2" />
          <TreeNode title="node2" key="0-0-3" />
        </TreeNode>
      </Tree>,
    );
    expect(container.querySelectorAll('.switcherIcon').length).toBe(1);
  });

  it('leaf nodes should render custom icons when provided', () => {
    const { container } = render(
      <Tree showLine={{ showLeafIcon: <i className="customLeafIcon" /> }} defaultExpandAll>
        <TreeNode icon="icon">
          <TreeNode title="node1" icon="icon" key="0-0-2" />
          <TreeNode title="node2" key="0-0-3" />
        </TreeNode>
      </Tree>,
    );
    expect(container.querySelectorAll('.customLeafIcon').length).toBe(2);
  });

  it('leaf nodes should render custom icons when provided as render function', () => {
    const { container } = render(
      <Tree showLine={{ showLeafIcon: () => <i className="customLeafIcon" /> }} defaultExpandAll>
        <TreeNode icon="icon">
          <TreeNode title="node1" icon="icon" key="0-0-2" />
          <TreeNode title="node2" key="0-0-3" />
        </TreeNode>
      </Tree>,
    );

    expect(container.querySelectorAll('.customLeafIcon').length).toBe(2);
  });

  it('leaf nodes should render custom icons when provided as string', async () => {
    render(
      <Tree showLine={{ showLeafIcon: 'customLeafIcon' }} defaultExpandAll>
        <TreeNode icon="icon">
          <TreeNode title="node1" icon="icon" key="0-0-2" />
          <TreeNode title="node2" key="0-0-3" />
        </TreeNode>
      </Tree>,
    );

    const customIcons = await screen.findAllByText('customLeafIcon');
    expect(customIcons).toHaveLength(2);
  });

  it('switcherIcon in Tree could be string', () => {
    const { asFragment } = render(
      <Tree switcherIcon="switcherIcon" defaultExpandAll>
        <TreeNode icon="icon">
          <TreeNode title="node1" icon="icon" key="0-0-2" />
          <TreeNode title="node2" key="0-0-3" />
        </TreeNode>
      </Tree>,
    );
    expect(asFragment().firstChild).toMatchSnapshot();
  });

  it('switcherIcon should be loading icon when loadData', () => {
    const onLoadData = () =>
      new Promise<void>((resolve) => {
        setTimeout(() => {
          resolve();
        }, 1000);
      });
    const { asFragment } = render(
      <Tree switcherIcon="switcherIcon" defaultExpandAll loadData={onLoadData}>
        <TreeNode icon="icon">
          <TreeNode title="node1" icon="icon" key="0-0-2" />
          <TreeNode title="node2" key="0-0-3" />
        </TreeNode>
      </Tree>,
    );
    expect(asFragment().firstChild).toMatchSnapshot();
  });

  it('support switcherLoadingIcon prop when loadData', () => {
    const onLoadData = () =>
      new Promise<void>((resolve) => {
        setTimeout(() => {
          resolve();
        }, 1000);
      });
    const { asFragment } = render(
      <Tree
        switcherIcon="switcherIcon"
        loadData={onLoadData}
        defaultExpandedKeys={['0-0-2', '0-0-3']}
        switcherLoadingIcon={<div>loading...</div>}
      >
        <TreeNode icon="icon">
          <TreeNode title="node1" icon="icon" key="0-0-2" />
          <TreeNode title="node2" key="0-0-3" />
        </TreeNode>
      </Tree>,
    );
    expect(asFragment().firstChild).toMatchSnapshot();
  });

  it('switcherIcon in Tree could be render prop function', () => {
    const { container } = render(
      <Tree
        defaultExpandAll
        switcherIcon={({ expanded }: AntTreeNodeProps) =>
          expanded ? <span className="open" /> : <span className="close" />
        }
      >
        <TreeNode icon="icon">
          <TreeNode title="node1" icon="icon" key="0-0-2" />
          <TreeNode title="node2" key="0-0-3" />
        </TreeNode>
      </Tree>,
    );
    expect(container.querySelectorAll('.open').length).toBe(1);
  });

  // https://github.com/ant-design/ant-design/issues/23261
  it('showLine is object type should render correctly', () => {
    const { asFragment } = render(
      <Tree showLine={{ showLeafIcon: false }} defaultExpandedKeys={['0-0-0']}>
        <TreeNode title="parent 1" key="0-0">
          <TreeNode title="parent 1-0" key="0-0-0">
            <TreeNode title="leaf" key="0-0-0-0" />
            <TreeNode title="leaf" key="0-0-0-1" />
            <TreeNode title="leaf" key="0-0-0-2" />
          </TreeNode>
          <TreeNode title="parent 1-1" key="0-0-1">
            <TreeNode title="leaf" key="0-0-1-0" />
          </TreeNode>
          <TreeNode title="parent 1-2" key="0-0-2">
            <TreeNode title="leaf" key="0-0-2-0" />
            <TreeNode title="leaf" key="0-0-2-1" />
          </TreeNode>
        </TreeNode>
      </Tree>,
    );
    expect(asFragment().firstChild).toMatchSnapshot();
  });

  describe('draggable', () => {
    const dragTreeData = [
      {
        title: 'bamboo',
        key: 'bamboo',
      },
    ];

    it('hide icon', () => {
      const { container } = render(<Tree treeData={dragTreeData} draggable={{ icon: false }} />);
      expect(container.querySelector('.anticon-holder')).toBeFalsy();
    });

    it('customize icon', () => {
      const { container } = render(
        <Tree treeData={dragTreeData} draggable={{ icon: <span className="little" /> }} />,
      );
      expect(container.querySelector('.little')).toBeTruthy();
    });

    it('nodeDraggable', () => {
      const nodeDraggable = jest.fn(() => false);
      render(<Tree treeData={dragTreeData} draggable={{ nodeDraggable }} />);
      expect(nodeDraggable).toHaveBeenCalledWith(dragTreeData[0]);
    });

    it('nodeDraggable func', () => {
      const nodeDraggable = jest.fn(() => false);
      render(<Tree treeData={dragTreeData} draggable={nodeDraggable} />);
      expect(nodeDraggable).toHaveBeenCalledWith(dragTreeData[0]);
    });
  });

  describe('hidden switcherIcon', () => {
    it('use `switcherIcon={() => null}`', () => {
      const { container } = render(
        <Tree defaultExpandAll switcherIcon={() => null}>
          <TreeNode icon="icon">
            <TreeNode title="node1" icon="icon" key="0-0-2" />
            <TreeNode title="node2" key="0-0-3" />
          </TreeNode>
        </Tree>,
      );
      container.querySelectorAll('.ant-tree-switcher').forEach((el) => {
        expect(el.children.length).toBe(0);
      });
    });
    it('use `switcherIcon={null}`', () => {
      const { container } = render(
        <Tree defaultExpandAll switcherIcon={null}>
          <TreeNode icon="icon">
            <TreeNode title="node1" icon="icon" key="0-0-2" />
            <TreeNode title="node2" key="0-0-3" />
          </TreeNode>
        </Tree>,
      );
      container.querySelectorAll('.ant-tree-switcher').forEach((el) => {
        expect(el.children.length).toBe(0);
      });
    });
  });
});
