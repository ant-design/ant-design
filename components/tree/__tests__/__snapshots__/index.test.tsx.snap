// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree icon and switcherIcon of Tree with showLine should render correctly 1`] = `
<div
  class="ant-tree ant-tree-show-line"
  role="tree"
>
  <div>
    <input
      aria-label="for screen reader"
      style="width: 0px; height: 0px; display: flex; overflow: hidden; opacity: 0; border: 0px; padding: 0px; margin: 0px;"
      tabindex="0"
      value=""
    />
  </div>
  <div
    aria-hidden="true"
    class="ant-tree-treenode"
    style="position: absolute; pointer-events: none; visibility: hidden; height: 0px; overflow: hidden; border: 0px; padding: 0px;"
  >
    <div
      class="ant-tree-indent"
    >
      <div
        class="ant-tree-indent-unit"
      />
    </div>
  </div>
  <div
    class="ant-tree-list"
    style="position: relative;"
  >
    <div
      class="ant-tree-list-holder"
    >
      <div>
        <div
          class="ant-tree-list-holder-inner"
          style="display: flex; flex-direction: column;"
        >
          <div
            aria-grabbed="false"
            class="ant-tree-treenode ant-tree-treenode-switcher-close"
            draggable="false"
          >
            <span
              aria-hidden="true"
              class="ant-tree-indent"
            />
            <span
              class="ant-tree-switcher ant-tree-switcher_close"
            >
              switcherIcon
            </span>
            <span
              class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-close"
              title="---"
            >
              <span
                class="ant-tree-iconEle ant-tree-icon__customize"
              >
                icon
              </span>
              <span
                class="ant-tree-title"
              >
                ---
              </span>
            </span>
          </div>
          <div
            aria-grabbed="false"
            class="ant-tree-treenode ant-tree-treenode-switcher-close"
            draggable="false"
          >
            <span
              aria-hidden="true"
              class="ant-tree-indent"
            />
            <span
              class="ant-tree-switcher ant-tree-switcher_close"
            >
              switcherIcon
            </span>
            <span
              class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-close"
              title="---"
            >
              <span
                class="ant-tree-iconEle ant-tree-icon__close"
              />
              <span
                class="ant-tree-title"
              >
                ---
              </span>
            </span>
          </div>
          <div
            aria-grabbed="false"
            class="ant-tree-treenode ant-tree-treenode-switcher-close ant-tree-treenode-leaf-last"
            draggable="false"
          >
            <span
              aria-hidden="true"
              class="ant-tree-indent"
            />
            <span
              class="ant-tree-switcher ant-tree-switcher_close"
            >
              <span
                aria-label="plus-square"
                class="anticon anticon-plus-square ant-tree-switcher-line-icon"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="plus-square"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M328 544h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8z"
                  />
                  <path
                    d="M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z"
                  />
                </svg>
              </span>
            </span>
            <span
              class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-close"
              title="---"
            >
              <span
                class="ant-tree-iconEle ant-tree-icon__close"
              />
              <span
                class="ant-tree-title"
              >
                ---
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Tree showLine is object type should render correctly 1`] = `
<div
  class="ant-tree ant-tree-icon-hide ant-tree-show-line"
  role="tree"
>
  <div>
    <input
      aria-label="for screen reader"
      style="width: 0px; height: 0px; display: flex; overflow: hidden; opacity: 0; border: 0px; padding: 0px; margin: 0px;"
      tabindex="0"
      value=""
    />
  </div>
  <div
    aria-hidden="true"
    class="ant-tree-treenode"
    style="position: absolute; pointer-events: none; visibility: hidden; height: 0px; overflow: hidden; border: 0px; padding: 0px;"
  >
    <div
      class="ant-tree-indent"
    >
      <div
        class="ant-tree-indent-unit"
      />
    </div>
  </div>
  <div
    class="ant-tree-list"
    style="position: relative;"
  >
    <div
      class="ant-tree-list-holder"
    >
      <div>
        <div
          class="ant-tree-list-holder-inner"
          style="display: flex; flex-direction: column;"
        >
          <div
            aria-grabbed="false"
            class="ant-tree-treenode ant-tree-treenode-switcher-open ant-tree-treenode-leaf-last"
            draggable="false"
          >
            <span
              aria-hidden="true"
              class="ant-tree-indent"
            />
            <span
              class="ant-tree-switcher ant-tree-switcher_open"
            >
              <span
                aria-label="minus-square"
                class="anticon anticon-minus-square ant-tree-switcher-line-icon"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="minus-square"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M328 544h368c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8z"
                  />
                  <path
                    d="M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z"
                  />
                </svg>
              </span>
            </span>
            <span
              class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-open"
              title="parent 1"
            >
              <span
                class="ant-tree-title"
              >
                parent 1
              </span>
            </span>
          </div>
          <div
            aria-grabbed="false"
            class="ant-tree-treenode ant-tree-treenode-switcher-open"
            draggable="false"
          >
            <span
              aria-hidden="true"
              class="ant-tree-indent"
            >
              <span
                class="ant-tree-indent-unit ant-tree-indent-unit-start ant-tree-indent-unit-end"
              />
            </span>
            <span
              class="ant-tree-switcher ant-tree-switcher_open"
            >
              <span
                aria-label="minus-square"
                class="anticon anticon-minus-square ant-tree-switcher-line-icon"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="minus-square"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M328 544h368c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8z"
                  />
                  <path
                    d="M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z"
                  />
                </svg>
              </span>
            </span>
            <span
              class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-open"
              title="parent 1-0"
            >
              <span
                class="ant-tree-title"
              >
                parent 1-0
              </span>
            </span>
          </div>
          <div
            aria-grabbed="false"
            class="ant-tree-treenode ant-tree-treenode-switcher-close"
            draggable="false"
          >
            <span
              aria-hidden="true"
              class="ant-tree-indent"
            >
              <span
                class="ant-tree-indent-unit ant-tree-indent-unit-start ant-tree-indent-unit-end"
              />
              <span
                class="ant-tree-indent-unit ant-tree-indent-unit-start"
              />
            </span>
            <span
              class="ant-tree-switcher ant-tree-switcher-noop"
            >
              <span
                class="ant-tree-switcher-leaf-line"
              />
            </span>
            <span
              class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-normal"
              title="leaf"
            >
              <span
                class="ant-tree-title"
              >
                leaf
              </span>
            </span>
          </div>
          <div
            aria-grabbed="false"
            class="ant-tree-treenode ant-tree-treenode-switcher-close"
            draggable="false"
          >
            <span
              aria-hidden="true"
              class="ant-tree-indent"
            >
              <span
                class="ant-tree-indent-unit ant-tree-indent-unit-start ant-tree-indent-unit-end"
              />
              <span
                class="ant-tree-indent-unit ant-tree-indent-unit-start"
              />
            </span>
            <span
              class="ant-tree-switcher ant-tree-switcher-noop"
            >
              <span
                class="ant-tree-switcher-leaf-line"
              />
            </span>
            <span
              class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-normal"
              title="leaf"
            >
              <span
                class="ant-tree-title"
              >
                leaf
              </span>
            </span>
          </div>
          <div
            aria-grabbed="false"
            class="ant-tree-treenode ant-tree-treenode-switcher-close ant-tree-treenode-leaf-last"
            draggable="false"
          >
            <span
              aria-hidden="true"
              class="ant-tree-indent"
            >
              <span
                class="ant-tree-indent-unit ant-tree-indent-unit-start ant-tree-indent-unit-end"
              />
              <span
                class="ant-tree-indent-unit ant-tree-indent-unit-start"
              />
            </span>
            <span
              class="ant-tree-switcher ant-tree-switcher-noop"
            >
              <span
                class="ant-tree-switcher-leaf-line"
              />
            </span>
            <span
              class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-normal"
              title="leaf"
            >
              <span
                class="ant-tree-title"
              >
                leaf
              </span>
            </span>
          </div>
          <div
            aria-grabbed="false"
            class="ant-tree-treenode ant-tree-treenode-switcher-close"
            draggable="false"
          >
            <span
              aria-hidden="true"
              class="ant-tree-indent"
            >
              <span
                class="ant-tree-indent-unit ant-tree-indent-unit-start ant-tree-indent-unit-end"
              />
            </span>
            <span
              class="ant-tree-switcher ant-tree-switcher_close"
            >
              <span
                aria-label="plus-square"
                class="anticon anticon-plus-square ant-tree-switcher-line-icon"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="plus-square"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M328 544h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8z"
                  />
                  <path
                    d="M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z"
                  />
                </svg>
              </span>
            </span>
            <span
              class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-close"
              title="parent 1-1"
            >
              <span
                class="ant-tree-title"
              >
                parent 1-1
              </span>
            </span>
          </div>
          <div
            aria-grabbed="false"
            class="ant-tree-treenode ant-tree-treenode-switcher-close ant-tree-treenode-leaf-last"
            draggable="false"
          >
            <span
              aria-hidden="true"
              class="ant-tree-indent"
            >
              <span
                class="ant-tree-indent-unit ant-tree-indent-unit-start ant-tree-indent-unit-end"
              />
            </span>
            <span
              class="ant-tree-switcher ant-tree-switcher_close"
            >
              <span
                aria-label="plus-square"
                class="anticon anticon-plus-square ant-tree-switcher-line-icon"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="plus-square"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M328 544h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8z"
                  />
                  <path
                    d="M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z"
                  />
                </svg>
              </span>
            </span>
            <span
              class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-close"
              title="parent 1-2"
            >
              <span
                class="ant-tree-title"
              >
                parent 1-2
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Tree support switcherLoadingIcon prop when loadData 1`] = `
<div
  class="ant-tree ant-tree-icon-hide"
  role="tree"
>
  <div>
    <input
      aria-label="for screen reader"
      style="width: 0px; height: 0px; display: flex; overflow: hidden; opacity: 0; border: 0px; padding: 0px; margin: 0px;"
      tabindex="0"
      value=""
    />
  </div>
  <div
    aria-hidden="true"
    class="ant-tree-treenode"
    style="position: absolute; pointer-events: none; visibility: hidden; height: 0px; overflow: hidden; border: 0px; padding: 0px;"
  >
    <div
      class="ant-tree-indent"
    >
      <div
        class="ant-tree-indent-unit"
      />
    </div>
  </div>
  <div
    class="ant-tree-list"
    style="position: relative;"
  >
    <div
      class="ant-tree-list-holder"
    >
      <div>
        <div
          class="ant-tree-list-holder-inner"
          style="display: flex; flex-direction: column;"
        >
          <div
            aria-grabbed="false"
            class="ant-tree-treenode ant-tree-treenode-switcher-open ant-tree-treenode-leaf-last"
            draggable="false"
          >
            <span
              aria-hidden="true"
              class="ant-tree-indent"
            />
            <span
              class="ant-tree-switcher ant-tree-switcher_open"
            >
              switcherIcon
            </span>
            <span
              class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-open"
              title="---"
            >
              <span
                class="ant-tree-title"
              >
                ---
              </span>
            </span>
          </div>
          <div
            aria-grabbed="false"
            class="ant-tree-treenode ant-tree-treenode-switcher-open ant-tree-treenode-loading"
            draggable="false"
          >
            <span
              aria-hidden="true"
              class="ant-tree-indent"
            >
              <span
                class="ant-tree-indent-unit ant-tree-indent-unit-start ant-tree-indent-unit-end"
              />
            </span>
            <span
              class="ant-tree-switcher ant-tree-switcher_open"
            >
              <div>
                loading...
              </div>
            </span>
            <span
              class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-open"
              title="node1"
            >
              <span
                class="ant-tree-iconEle ant-tree-icon__open ant-tree-icon_loading"
              />
              <span
                class="ant-tree-title"
              >
                node1
              </span>
            </span>
          </div>
          <div
            aria-grabbed="false"
            class="ant-tree-treenode ant-tree-treenode-switcher-open ant-tree-treenode-loading ant-tree-treenode-leaf-last"
            draggable="false"
          >
            <span
              aria-hidden="true"
              class="ant-tree-indent"
            >
              <span
                class="ant-tree-indent-unit ant-tree-indent-unit-start ant-tree-indent-unit-end"
              />
            </span>
            <span
              class="ant-tree-switcher ant-tree-switcher_open"
            >
              <div>
                loading...
              </div>
            </span>
            <span
              class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-open"
              title="node2"
            >
              <span
                class="ant-tree-iconEle ant-tree-icon__open ant-tree-icon_loading"
              />
              <span
                class="ant-tree-title"
              >
                node2
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Tree switcherIcon in Tree could be string 1`] = `
<div
  class="ant-tree ant-tree-icon-hide"
  role="tree"
>
  <div>
    <input
      aria-label="for screen reader"
      style="width: 0px; height: 0px; display: flex; overflow: hidden; opacity: 0; border: 0px; padding: 0px; margin: 0px;"
      tabindex="0"
      value=""
    />
  </div>
  <div
    aria-hidden="true"
    class="ant-tree-treenode"
    style="position: absolute; pointer-events: none; visibility: hidden; height: 0px; overflow: hidden; border: 0px; padding: 0px;"
  >
    <div
      class="ant-tree-indent"
    >
      <div
        class="ant-tree-indent-unit"
      />
    </div>
  </div>
  <div
    class="ant-tree-list"
    style="position: relative;"
  >
    <div
      class="ant-tree-list-holder"
    >
      <div>
        <div
          class="ant-tree-list-holder-inner"
          style="display: flex; flex-direction: column;"
        >
          <div
            aria-grabbed="false"
            class="ant-tree-treenode ant-tree-treenode-switcher-open ant-tree-treenode-leaf-last"
            draggable="false"
          >
            <span
              aria-hidden="true"
              class="ant-tree-indent"
            />
            <span
              class="ant-tree-switcher ant-tree-switcher_open"
            >
              switcherIcon
            </span>
            <span
              class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-open"
              title="---"
            >
              <span
                class="ant-tree-title"
              >
                ---
              </span>
            </span>
          </div>
          <div
            aria-grabbed="false"
            class="ant-tree-treenode ant-tree-treenode-switcher-close"
            draggable="false"
          >
            <span
              aria-hidden="true"
              class="ant-tree-indent"
            >
              <span
                class="ant-tree-indent-unit ant-tree-indent-unit-start ant-tree-indent-unit-end"
              />
            </span>
            <span
              class="ant-tree-switcher ant-tree-switcher-noop"
            />
            <span
              class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-normal"
              title="node1"
            >
              <span
                class="ant-tree-title"
              >
                node1
              </span>
            </span>
          </div>
          <div
            aria-grabbed="false"
            class="ant-tree-treenode ant-tree-treenode-switcher-close ant-tree-treenode-leaf-last"
            draggable="false"
          >
            <span
              aria-hidden="true"
              class="ant-tree-indent"
            >
              <span
                class="ant-tree-indent-unit ant-tree-indent-unit-start ant-tree-indent-unit-end"
              />
            </span>
            <span
              class="ant-tree-switcher ant-tree-switcher-noop"
            />
            <span
              class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-normal"
              title="node2"
            >
              <span
                class="ant-tree-title"
              >
                node2
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Tree switcherIcon should be loading icon when loadData 1`] = `
<div
  class="ant-tree ant-tree-icon-hide"
  role="tree"
>
  <div>
    <input
      aria-label="for screen reader"
      style="width: 0px; height: 0px; display: flex; overflow: hidden; opacity: 0; border: 0px; padding: 0px; margin: 0px;"
      tabindex="0"
      value=""
    />
  </div>
  <div
    aria-hidden="true"
    class="ant-tree-treenode"
    style="position: absolute; pointer-events: none; visibility: hidden; height: 0px; overflow: hidden; border: 0px; padding: 0px;"
  >
    <div
      class="ant-tree-indent"
    >
      <div
        class="ant-tree-indent-unit"
      />
    </div>
  </div>
  <div
    class="ant-tree-list"
    style="position: relative;"
  >
    <div
      class="ant-tree-list-holder"
    >
      <div>
        <div
          class="ant-tree-list-holder-inner"
          style="display: flex; flex-direction: column;"
        >
          <div
            aria-grabbed="false"
            class="ant-tree-treenode ant-tree-treenode-switcher-open ant-tree-treenode-leaf-last"
            draggable="false"
          >
            <span
              aria-hidden="true"
              class="ant-tree-indent"
            />
            <span
              class="ant-tree-switcher ant-tree-switcher_open"
            >
              switcherIcon
            </span>
            <span
              class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-open"
              title="---"
            >
              <span
                class="ant-tree-title"
              >
                ---
              </span>
            </span>
          </div>
          <div
            aria-grabbed="false"
            class="ant-tree-treenode ant-tree-treenode-switcher-close"
            draggable="false"
          >
            <span
              aria-hidden="true"
              class="ant-tree-indent"
            >
              <span
                class="ant-tree-indent-unit ant-tree-indent-unit-start ant-tree-indent-unit-end"
              />
            </span>
            <span
              class="ant-tree-switcher ant-tree-switcher_close"
            >
              switcherIcon
            </span>
            <span
              class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-close"
              title="node1"
            >
              <span
                class="ant-tree-title"
              >
                node1
              </span>
            </span>
          </div>
          <div
            aria-grabbed="false"
            class="ant-tree-treenode ant-tree-treenode-switcher-close ant-tree-treenode-leaf-last"
            draggable="false"
          >
            <span
              aria-hidden="true"
              class="ant-tree-indent"
            >
              <span
                class="ant-tree-indent-unit ant-tree-indent-unit-start ant-tree-indent-unit-end"
              />
            </span>
            <span
              class="ant-tree-switcher ant-tree-switcher_close"
            >
              switcherIcon
            </span>
            <span
              class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-close"
              title="node2"
            >
              <span
                class="ant-tree-title"
              >
                node2
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
